bpo-34060: Report system load when running test suite for Windows

ammaraskar opened the pull request:
This is (mostly) a pure Python implementation of the other PR. It leverages the `typeperf` command which monitors performance counters and outputs them at a given interval. So every 5 seconds, `typeperf` can output the processor queue length into stdout.

`subprocess.stdout.readline` is a blocking call. Using a thread seemed like an obvious solution, but we can't achieve this with multiprocessing or a thread, because like Victor speculated in the previous bug report on this, it conflicts with `test_multiprocessing` and `test_threading`. Hence, I opted to use the asynchronous/overlapped IO API which was designed for `async`. Most of the diff actually just pertains to using this rather low level API.

This is almost a pure python implementation but there was one edge case where this would fail. Namely, when the python interpreter running the test suite crashes, this leaves an orphaned `typeperf` process running which refuses to die. This means that when the test suite is run with `-j x` and this situation happens:

```
python -m test -j2
├── python <test runner>
│   └── typeperf.exe
└── python *CRASHED*
    └── typeperf.exe
```

The big test coordinating python process will wait forever on the crashed python and consequently `typeperf` to terminate, which just doesn't happen by default in Windows. After reading up on the APIs, the right way to fix this is by using a Job Object to ask the [OS to kill the child when the parent dies](https://stackoverflow.com/questions/53208/how-do-i-automatically-destroy-child-processes-in-windows). Hence, there is a change in `_winapi` to make this happen. Unlike the last PR, this API is actually reusable and fit to be exposed to the public. It could even allow implementing things like [bpo-5115](https://bugs.python.org/issue5115) to be a lot easier.

<!-- issue-number: bpo-34060 -->
https://bugs.python.org/issue34060
<!-- /issue-number -->

zooba commented on Modules/_winapi.c:
```diff
@@ -1699,6 +1699,69 @@ _winapi_GetFileType_impl(PyObject *module, HANDLE handle)
+                             Py_UNICODE *job_name)
+/*[clinic end generated code: output=ddb90fccf4f363b9 input=f1350a07abdab61c]*/
+{
+    HANDLE job_handle = CreateJobObjectW(NULL, job_name);
```
We should surround the non-Python parts of this function with `_Py_BEGIN_ALLOW_THREADS` and `_Py_END_ALLOW_THREADS` to release the GIL. Similarly for the Assign function.

zooba reviewed (approved):
I like this! And I'll be happy to have support for job objects in there too :)

Looking at the test runs, the numbers seem to be consistent with other platforms, but since I don't have as good a feel for what to expect here, I'd like someone else who's been involved in this or the previous PR to sign off as well.

zooba commented on Lib/test/libregrtest/main.py:
```diff
@@ -100,6 +101,13 @@ def __init__(self):
+            self.getloadavg = lambda: os.getloadavg()[0]
+        elif sys.platform == 'win32':
+            self.load_tracker = WindowsLoadTracker()
+            self.getloadavg = lambda: self.load_tracker.getloadavg()
```
This can just be `self.getloadavg = self.load_tracker.getloadavg`

zooba replied:
And since there are no later references to `self.load_tracker`, that doesn't need to be kept on the object (the references via the method will keep it from being deallocated)

ammaraskar replied:
Done, thanks :)

eryksun commented on Modules/_winapi.c:
```diff
@@ -1699,6 +1699,80 @@ _winapi_GetFileType_impl(PyObject *module, HANDLE handle)
+
+    if (job_handle == NULL) {
+        PyErr_SetFromWindowsErr(GetLastError());
+        return NULL;
```
`HANDLE_return_converter` requires `INVALID_HANDLE_VALUE` for an error.

ammaraskar replied:
Good catch, thanks.

eryksun commented on Lib/test/libregrtest/utils.py:
```diff
@@ -54,3 +56,102 @@ def printlist(x, width=70, indent=4, file=None):
+
+        # Spawn off the load monitor
+        command = ['typeperf', COUNTER_NAME, '-si', str(SAMPLING_INTERVAL)]
+        self.p = subprocess.Popen(command, stdout=command_stdout)
```
Python still supports Windows 7, which only allows a process to be assigned to one job at a time. If the OS version is prior to NT 6.2 (i.e. `sys.getwindowsversion() < (6, 2)`), use the creation flag 
`CREATE_BREAKAWAY_FROM_JOB` (0x01000000) to try to break the child process out of the current job.

ammaraskar replied:
Aah thank you, I've been testing on Windows 10 and didn't read the documentation carefully enough, is there any particular reason to scope it down to <NT 6.2? Will there be a problem with just using `CREATE_BREAKAWAY_FROM_JOB` as the `creationflags` all the time?

eryksun replied:
There's no reason to break the child out of the job hierarchy in Windows 8+, especially since jobs are used more frequently (e.g. to implement silos, such as for Windows containers). 

Also, trying to break away from the parent job complicates creating the process. I forgot to mention that `Popen` will fail with a `PermissionError` if the child isn't allowed to break away. This can be handled by retrying without `CREATE_BREAKAWAY_FROM_JOB`. In this case you could skip calling `AssignProcessToJobObject`. For example (untested):

    CREATE_BREAKAWAY_FROM_JOB = 0x01000000

    cflags = 0 if sys.getwindowsversion() >= (6, 2) else CREATE_BREAKAWAY_FROM_JOB
    assign_to_job = True

    try:
        self.p = subprocess.Popen(command, stdout=command_stdout,
                    creationflags=cflags)
    except PermissionError:
        if cflags == 0:
            raise
        self.p = subprocess.Popen(command, stdout=command_stdout)
        assign_to_job = False

    if assign_to_job:
        _winapi.AssignProcessToJobObject(job_group, self.p._handle)

eryksun commented on Lib/test/libregrtest/utils.py:
```diff
@@ -54,3 +56,102 @@ def printlist(x, width=70, indent=4, file=None):
+        self.p = subprocess.Popen(command, stdout=command_stdout)
+
+        # Attach it to our job group
+        _winapi.AssignProcessToJobObject(job_group, self.p._handle)
```
Call this in a try/except that handles failure either by ignoring it or with a warning. `AssignProcessToJobObject` will fail in Windows 7 if the parent process (i.e. the current Python process) is in a job that doesn't allow child processes to break away.

ammaraskar replied:
Thanks, I'll issue a warning since this to handle a rather rare case of the interpreter actually crashing in `-jN` mode.

eryksun commented on Modules/_winapi.c:
```diff
@@ -1699,6 +1699,80 @@ _winapi_GetFileType_impl(PyObject *module, HANDLE handle)
+    // we only support a subset of all the limits possible, can be expanded
+    // in the future if needed.
+    limit.BasicLimitInformation.LimitFlags =
+        (DWORD) getulong(limit_info, "LimitFlags");
```
Maybe this should reflect the WinAPI structure, i.e. the `BasicLimitInformation` field. Also, it would be nice to have PyWin32 interoperability, which does the latter and also uses dicts instead of simple namespaces.

ammaraskar replied:
Hmm, so I based this on `CreateProcess` which also uses attributes in a passed in object for this information. What do you mean by reflect the structure? Do you mean support all the fields for `BasicLimitInformation`?

ammaraskar replied:
For reference, here is how subprocess passes uses it with `CreateProcess` https://github.com/python/cpython/blob/06ca3f0c09d017b9d741553818459cca2d5da587/Lib/subprocess.py#L129-L137

I just used simple namespace because I was being lazy

eryksun replied:
Sometimes a dict is used, e.g. `STARTUPINFO` uses a dict for the new `lpAttributeList` support in 3.7. `win32job` uses dicts, but it's fine to stick with namespaces instead. I still would rather `LimitFlags` be set in a `BasicLimitInformation` attribute instead of at the top level. _winapi is low-level and undocumented, so it's best if it mirrors the actual API, if it's not overly awkward. I didn't meant to support all fields, however. _winapi gets extended only as required.

ammaraskar committed:
bpo-34060: Report system load when running test suite for Windows

While Windows exposes the system processor queue length: the raw value
used for load calculations on Unix systems, it does not provide an API
to access the averaged value. Hence to calculate the load we must track
and average it ourselves. We can't use multiprocessing or a thread to
read it in the background while the tests run since using those would
conflict with test_multiprocessing and test_xxsubprocess.

Thus, we use Window's asynchronous IO API to run the tracker in the
background with it sampling at the correct rate. When we wish to access
the load we check to see if there's new data on the stream, if there is,
we update our load values.

vstinner commented on Lib/test/libregrtest/main.py:
```diff
@@ -533,6 +534,13 @@ def main(self, tests=None, **kwargs):
+        self.getloadavg = None
+        if hasattr(os, 'getloadavg'):
+            self.getloadavg = lambda: os.getloadavg()[0]
+        elif sys.platform == 'win32' and (self.ns.slaveargs is None):
```
I would prefer to import the class there.

ammaraskar replied:
Done.

vstinner commented on Lib/test/libregrtest/main.py:
```diff
@@ -533,6 +534,13 @@ def main(self, tests=None, **kwargs):
 
+        self.getloadavg = None
+        if hasattr(os, 'getloadavg'):
+            self.getloadavg = lambda: os.getloadavg()[0]
```
I dislike lambda. Would you mind to define a method here using "def" ?

ammaraskar replied:
I prefer lambda for this sort of short stuff but done

vstinner commented on Lib/test/libregrtest/utils.py:
```diff
@@ -54,3 +58,94 @@ def printlist(x, width=70, indent=4, file=None):
                         initial_indent=blanks, subsequent_indent=blanks),
           file=file)
+
+BUFSIZE = 8192
```
Please put the new code in a new file, like libregrtest/winloadavg.py.

ammaraskar replied:
Done

vstinner commented on Lib/test/libregrtest/utils.py:
```diff
@@ -54,3 +58,94 @@ def printlist(x, width=70, indent=4, file=None):
+COUNTER_NAME = r'\System\Processor Queue Length'
+
+"""
+This class asynchronously interacts with the `typeperf` command to read
```
Please put the docstring inside the class.

vstinner commented on Lib/test/libregrtest/utils.py:
```diff
@@ -54,3 +58,94 @@ def printlist(x, width=70, indent=4, file=None):
+
+    def __del__(self):
+        self.p.kill()
+        self.p.wait()
```
Please don't use __del__() but add a close() method for example.

ammaraskar replied:
Hmm, are you sure about this? It would require us to keep a reference to `load_tracker` and it needs an:
```python
if self.load_tracker:
    self.load_tracker.stop()
```
because we don't have a load tracker to stop on non windows systems.

vstinner commented on Lib/test/libregrtest/utils.py:
```diff
@@ -54,3 +58,94 @@ def printlist(x, width=70, indent=4, file=None):
+            return self.load
+
+        # Process the backlog of load values
+        for line in typeperf_output.splitlines():
```
What if the last line is incomplete: doesn't end with a newline character? Maybe you should put it back into a buffer, and concatenate it to the output, next time. Maybe use .splitlines(True) to check if there is a newline character?

ammaraskar replied:
I don't think its worth adding the extra complexity to handle this. Worst case is we miss a single point or two of data and the load number is slightly off.

vstinner commented on Lib/test/libregrtest/utils.py:
```diff
@@ -54,3 +58,94 @@ def printlist(x, width=70, indent=4, file=None):
+
+BUFSIZE = 8192
+LOAD_FACTOR_1 = 0.9200444146293232478931553241
+SAMPLING_INTERVAL = 5
```
Please add a comment to document the unit (seconds, no?).

vstinner commented on Lib/test/libregrtest/utils.py:
```diff
@@ -54,3 +58,94 @@ def printlist(x, width=70, indent=4, file=None):
+        # Close our copy of the write end of the pipe
+        os.close(command_stdout)
+
+    def read_output(self):
```
It's possible that this function is only called every 5 minutes. I expect a lot of output in this case.

Why not running this function inside a thread?

ammaraskar replied:
This failure shows up when running in a thread: https://github.com/python/cpython/pull/8357#issuecomment-406771751

Even at 5 minutes, that's 60 points of data. Even if the typeperf command puts out 100 bytes of output per line its not enough to saturate the buffer. And even if it does, those data points will get picked up eventually by the next call.

vstinner commented on Lib/test/libregrtest/main.py:
```diff
@@ -533,6 +534,13 @@ def main(self, tests=None, **kwargs):
+            self.getloadavg = lambda: os.getloadavg()[0]
+        elif sys.platform == 'win32' and (self.ns.slaveargs is None):
+            load_tracker = WindowsLoadTracker()
+            self.getloadavg = load_tracker.getloadavg
```
I seems like you spawn a subprocess even in test worker processes, whereas you wrote that it's useless. This code should be moved after handling slaveargs, no? Maybe move the slaveargs handling code earlier?

ammaraskar replied:
Not sure I follow, we already know if the runner is a slave based on the parsing of the arguments. The arguments have been parsed by this point, so this should be fine. I also tested and under task manager only one typeperf instance shows up under the main python process unlike before.

ammaraskar committed:
Move windows specific code to its own file

ammaraskar committed:
Move imports to top of file

ammaraskar committed:
Add comment explaining check in libregrtest
